internal/component/discovery/relabel/relabel.go

```diff
@@ -82,24 +79,23 @@ func (c *Component) Run(ctx context.Context) error {
 
 // Update implements component.Component.
 func (c *Component) Update(args component.Arguments) error {
-	c.mut.Lock()
-	defer c.mut.Unlock()
-
 	newArgs := args.(Arguments)
 
 	targets := make([]discovery.Target, 0, len(newArgs.Targets))
-	relabelConfigs := alloy_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelConfigs)
-	c.rcs = relabelConfigs
 
 	for _, t := range newArgs.Targets {
-		lset := componentMapToPromLabels(t)
-		relabelled, keep := relabel.Process(lset, relabelConfigs...)
+		var (
+			relabelled discovery.Target
+			builder    = discovery.NewTargetBuilderFrom(t)
```

Comment:
This builder should/could be reused to avoid extra allocations

Reply:
Not sure it should be since the content of the builder is mutated on each call to ProcessBuilder, if we can make it safe then seems reasonable.

Reply:
Since the current profiles show good performance gains here already, I would like to leave this optimisation for future PRs after we have more data. I would like to consider an option of using an object pool for the couple of maps that the builder uses underneath as an alternative and compare the two options.